### PR TITLE
Xsoar mirroring bug

### PR DESCRIPTION
**Pull request by rshunim**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/XSUP-20748

## Description
Added a parameter 'Mirror Playbook ID' which allows to choose if mirroring the playbookId or not.
With this fix, the custom playbook on the remote side will not be transferred, so it will run according to the incident type's default playbook on the other side.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/README.md [hunk 1]
```diff
@@ -1,7 +1,6 @@
 Allows mirroring of XSOAR incidents between different Cortex XSOAR tenants.
 
 This integration was integrated and tested with version 6.0 of XSOAR
-Mirrored incidents include the playbook ID. The receiving side will attempt to run a playbook with a matching ID, if one exists locally.
```
Comment (yuvalbenshalom) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/README.md:4]: why did you remove this? the default behavior should still be that.

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.yml [hunk 2]
```diff
@@ -87,6 +87,12 @@ configuration:
   name: debug_mode
   type: 8
   required: false
+- display: Drop Playbook ID
+  name: drop_playbookId
+  defaultvalue: "false"
+  type: 8
+  required: false
+  additionalinfo: 'Should the integration mirror the playbook ID field to the remote system? Note: Mirroring the playbook ID means that the remote system must be running the same playbook as the source system.'
```
Comment (yuvalbenshalom) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.yml:95]: please rephrase this without "?"

Comment (yuvalbenshalom) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.yml:95]: please add documentation in the readme that explains the case that this is used.

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
see my comments

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py [hunk 3]
```diff
@@ -131,6 +137,52 @@ def test_fetch_incidents(mocker):
...
+    When:
+        - Running the fetch_incidents and getting this incident.
+
+    Then:
+        - Ensure the incident result contains playbookId field as expected.
+    """
+    mocker.patch.object(Client, 'search_incidents', return_value=INCIDENTS_MIRRORING_PLAYBOOK_ID)
+
+    first_fetch = dateparser.parse('3 days').strftime(XSOAR_DATE_FORMAT)
+    client = Client("")
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py:154]:
```suggestion
    client = Client("dummy token")
```
so future maintainers know what this empty string means

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py [hunk 4]
```diff
@@ -131,6 +137,52 @@ def test_fetch_incidents(mocker):
     assert dateparser.parse(next_run['last_fetch']) == dateparser.parse(INCIDENTS[-1]['created']) + timedelta(milliseconds=1)
 
 
+def test_fetch_incidents_mirroring_playbook_id(mocker):
+    """
+    Given:
+        - List of incident.
+
+    When:
+        - Running the fetch_incidents and getting this incident.
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py:146]:
```suggestion
        - Running the fetch_incidents and getting this incident, with the *implicit* default `drop_playbook_id = False`. 
```

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py [hunk 5]
```diff
@@ -131,6 +137,52 @@ def test_fetch_incidents(mocker):
...
+
+    Then:
+        - Ensure the incident result does not contain playbookId field as expected.
+    """
+    mocker.patch.object(Client, 'search_incidents', return_value=INCIDENTS_MIRRORING_PLAYBOOK_ID)
+
+    first_fetch = dateparser.parse('3 days').strftime(XSOAR_DATE_FORMAT)
+    client = Client("")
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py:177]:
```suggestion
    client = Client("dummy token")
```

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py [hunk 6]
```diff
@@ -131,6 +137,52 @@ def test_fetch_incidents(mocker):
...
+
+    next_run, incidents_result = fetch_incidents(client=client, max_results=3, last_run={}, first_fetch_time=first_fetch,
+                                                 query='', mirror_direction='None', mirror_tag=[], drop_playbookid=True)
+
+    assert len(incidents_result) == 1
+    assert "playbookId" not in incidents_result[0]
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py:183]: please parametrize the param `drop_playbookid` (once `True`, once `False`), then change to  
```python
    assert "playbookId" incidents_result[0] == not drop_playbookid
```

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py [hunk 7]
```diff
@@ -131,6 +137,52 @@ def test_fetch_incidents(mocker):
...
+def test_fetch_incidents_drop_playbook_id(mocker):
+    """
+    Given:
+        - List of incident.
+
+    When:
+        - Running the fetch_incidents and getting this incident.
+
+    Then:
+        - Ensure the incident result does not contain playbookId field as expected.
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py:172]:
```suggestion
        - Ensure the incident result does not contain playbookId field when `drop_playbookid` is True.
```

File: Packs/XSOARmirroring/ReleaseNotes/2_0_12.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### XSOAR Mirroring
+- Added a parameter 'Mirror Playbook ID' which allows to choose if mirroring the playbookId or not.
```
Comment (dorschw) [inline Packs/XSOARmirroring/ReleaseNotes/2_0_12.md:4]:
```suggestion
- Added the boolean `Drop Playbook ID` parameter. When set to true, the receiving instance will run the default playbook matching the incident type (if one exists).
```

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.py [hunk 9]
```diff
@@ -279,6 +283,9 @@ def fetch_incidents(client: Client, max_results: int, last_run: Dict[str, Union[
     """
 
     last_fetch = last_run.get('last_fetch')
+    # Drop playbookId field if configured.
+    if drop_playbookid and 'playbookId' in FIELDS_TO_COPY_FROM_REMOTE_INCIDENT:
+        FIELDS_TO_COPY_FROM_REMOTE_INCIDENT.remove('playbookId')
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.py:288]: Let's pass the parameter directly into `fetch_incidents`, and have it not copy this specific field, instead of changing the list of fields (which may cause trouble).

**Review by dorschw — COMMENTED** [review]
Very nice, please see the comments and re-request review.

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/README.md [hunk 10]
```diff
@@ -2,6 +2,7 @@ Allows mirroring of XSOAR incidents between different Cortex XSOAR tenants.
 
 This integration was integrated and tested with version 6.0 of XSOAR
 Mirrored incidents include the playbook ID. The receiving side will attempt to run a playbook with a matching ID, if one exists locally.
+The 'Drop Playbook ID' parameter must be selected when you want the source side to run according to its default playbook.
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/README.md:5]:
```suggestion
To have the machine run the default playbook for the mirrored incident, set the `Drop Playbook ID` to `true`. Otherwise (default), the machine will attempt to run a playbook whose id matches the `playbookId` field in the mirrored incident.
```

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.py [hunk 11]
```diff
@@ -790,7 +800,8 @@ def main() -> None:
                     first_fetch_time=first_fetch_time,
                     query=query,
                     mirror_direction=demisto.params().get('mirror_direction'),
-                    mirror_tag=list(mirror_tags)
+                    mirror_tag=list(mirror_tags),
+                    drop_playbook_id=demisto.params().get('drop_playbook_id')
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.py:804]:
```suggestion
                    drop_playbook_id=demisto.params().get('drop_playbook_id'),
```

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.py [hunk 12]
```diff
@@ -803,7 +814,8 @@ def main() -> None:
                 first_fetch_time=first_fetch_time,
                 query=query,
                 mirror_direction=demisto.params().get('mirror_direction'),
-                mirror_tag=list(mirror_tags)
+                mirror_tag=list(mirror_tags),
+                drop_playbook_id=demisto.params().get('drop_playbook_id')
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.py:818]:
```suggestion
                drop_playbook_id=demisto.params().get('drop_playbook_id'),
```

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py [hunk 13]
```diff
@@ -131,6 +138,31 @@ def test_fetch_incidents(mocker):
     assert dateparser.parse(next_run['last_fetch']) == dateparser.parse(INCIDENTS[-1]['created']) + timedelta(milliseconds=1)
 
 
+@pytest.mark.parametrize('drop_playbook_id', [False, True])
+def test_fetch_incidents_mirroring_or_dropping_playbook_id(mocker, drop_playbook_id: bool):
+    """
+    Given:
+        - List of incident.
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring_test.py:145]:
```suggestion
        - a list of incidents.
```

**Review by dorschw — APPROVED** [review]

File: Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.py [hunk 14]
```diff
@@ -790,7 +800,8 @@ def main() -> None:
                     first_fetch_time=first_fetch_time,
                     query=query,
                     mirror_direction=demisto.params().get('mirror_direction'),
-                    mirror_tag=list(mirror_tags)
+                    mirror_tag=list(mirror_tags),
+                    drop_playbook_id=demisto.params().get('drop_playbook_id'),
```
Comment (dorschw) [inline Packs/XSOARmirroring/Integrations/XSOARmirroring/XSOARmirroring.py:804]:
```suggestion
                    drop_playbook_id=demisto.params().get('drop_playbook_id', False),
```

File: Packs/XSOARmirroring/ReleaseNotes/2_0_12.md [hunk 15]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### XSOAR Mirroring
+- Added the boolean `Drop Playbook ID` parameter. When set to true, the receiving instance will run the default playbook matching the incident type (if one exists).
```
Comment (dorschw) [inline Packs/XSOARmirroring/ReleaseNotes/2_0_12.md:4]:
```suggestion
- Added the boolean `Mirror Playbook ID` parameter. When set to `false`, the receiving instance will run the default playbook matching the incident type (if one exists). Default is `true`, preseving backwards compatibility.
```

File: Packs/XSOARmirroring/ReleaseNotes/2_0_12.md [hunk 16]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### XSOAR Mirroring
+- Added the boolean `Mirror Playbook ID` parameter. When set to `false`, the receiving instance will run the default playbook matching the incident type (if one exists). Default is `true`, preseving backwards compatibility.
```
Comment (dorschw) [inline Packs/XSOARmirroring/ReleaseNotes/2_0_12.md:4]:
```suggestion
- Added the boolean `Mirror Playbook ID` parameter. Default is `true`, preseving backwards compatibility. When set to `false`, the receiving instance will run the default playbook matching the incident type (if one exists). 
```

**Review by yuvalbenshalom — APPROVED** [review]
